### PR TITLE
EIP-1344: Update uint256 field size to uint64

### PR DESCRIPTION
**Pull request by fubuloubu**
Per Discussion: https://ethereum-magicians.org/t/eip-1344-add-chain-id-opcode/1131/109

### CODE REVIEW

File: EIPS/eip-1344.md [hunk 1]
```diff
@@ -17,10 +17,12 @@ This EIP adds an opcode that returns the current chain's EIP-155 unique identifi
...
 
 ## Specification
-Adds a new opcode `CHAINID` at 0x46, which uses 0 stack arguments. It pushes the current chain ID onto the stack. Chain ID is a 256-bit value. The operation costs `G_base` to execute.
+Adds a new opcode `CHAINID` at 0x46, which uses 0 stack arguments. It pushes the current chain ID onto the stack. Chain ID is a 64-bit value. The operation costs `G_base` to execute.
 
 The value of the current chain ID is obtained from the chain ID configuration, which should match the EIP-155 unique identifier a client will accept from incoming transactions. Please note that per EIP-155, it is not *required* that a transaction have an EIP-155 unique identifier, but in that scenario this opcode will still return the configured chain ID and not a default.
 
+**NOTE**: A 64-bit value was chosen as a reasonably-sized value that works with the EIP-155 formatting for the `v` field (where `v = CHAIN_ID * 2 + 36`). The max value that can be used for chain ID is `MAX_CHAIN_ID = MAX_UNIT256 / 2 - 36` in order to have the math fit inside the RLP encoding for the `v` field in a transaction (as is currently expected by all production RLP encoders).
```
Comment (axic) [inline EIPS/eip-1344.md]: Typo: `MAX_UNIT256` -> `MAX_UINT256`.

Comment (fubuloubu) [reply EIPS/eip-1344.md]:
```suggestion
**NOTE**: A 64-bit value was chosen as a reasonably-sized value that works with the EIP-155 formatting for the `v` field (where `v = CHAIN_ID * 2 + 36`). The max value that can be used for chain ID is `MAX_CHAIN_ID = MAX_UINT256 / 2 - 36` in order to have the math fit inside the RLP encoding for the `v` field in a transaction (as is currently expected by all production RLP encoders).
```

**Comment by axic** [comment]
cc @holiman @shemnon @winsvega who are involved with test generation

**Comment by winsvega** [comment]
Chainid tests here:

https://github.com/ethereum/tests/pull/627

**Comment by chfast** [comment]
When merge?

**Comment by fubuloubu** [comment]
I think this is ready to merge, although I'm not sure if every client lead is aware of this change.

cc @sorpaas @karalabe

**Comment by shemnon** [comment]
So how would we do a reference test to validate this?  or is it validatable via reference test?

**Comment by fubuloubu** [comment]
Since it is possible to sign with with a chain ID > 64 bits (up to a value of ~255 bits IIRC), you could have an EIP-155 transaction validation that uses 65 bits and it should fail validation in some way. The opcode should also revert if the value of chain ID is > 64 bits.

**Comment by holiman** [comment]
> The opcode should also revert if the value of chain ID is > 64 bits.

Nope. No revertals for this opcode. It simply should not be possible to configure the node with a chainid above `64` bits. 

> you could have an EIP-155 transaction validation that uses 65 bits and it should fail validation in some way

So yes, that would obviously never equal the `64`-bit chainid

**Comment by axic** [comment]
To be fair this isn't something proposed last week, but has been under discussion since chaind was proposed and the current discussion [stems from a month ago](https://ethereum-magicians.org/t/eip-1344-add-chain-id-opcode/1131/108).

The problem with 256-bit chainid is that clients won't be able to handle it, unless their transaction encoding works on unrestricted bignums (and not just 256-bit bignums).

I really feel like this hardfork is rushed for no reason and dismissing every useful proposal just to be "done with the hard fork".

**Comment by karalabe** [comment]
> The problem with 256-bit chainid is that clients won't be able to handle it, unless their transaction encoding works on unrestricted bignums (and not just 256-bit bignums).

They need to handle that even now, since EIP155 already introduced this. So this PR limiting it to 64 bits makes it a weird cornercase that's not in line with EIP155. And this EIP not fixing EIP155 means that the whole fix is moot, because clients still need to handle large V-s properly.

**Comment by fubuloubu** [comment]
@karalabe:
1. This PR doesn't exist as a change to EIP-155, more a clarification. The change is to EIP-1344 because it had spelt out an exact data size for the EVM to return in it's errata. I do agree however that EIP-155 implementation is affected, but this should only make it clearer for clients because EIP-155 makes no considerations for size, even having an arbitrarily large size (except through the lack of an explanation).
2. At all tables of Chain IDs we have seen, no network has picked a value larger than 64 bits. It should therefore be safe to cap at this amount.
3. 63 bits could be better than 64 bits, because the math would fit into the smaller 64-bit integer. As it is, since unrestricted size on this parameter was not a clear assumption in EIP-155, there could be an overflow/encoding vulnerability in how different clients handle EIP-155. Because there was no expressed limit, there was never any testing done for much larger values either.

**Comment by tjayrush** [comment]
I am an un-involved observer, I have nothing to do with the hard-fork planning nor this EIP, and perhaps no claim to make the following comment, but I do have 39.25 years of experience with software development, so I'll go for it.

There are a few things that are quite obvious: (1) if 63 bits would be safer, then it should be 63 bits; (2) the spec should be explicit about this size of the item; (3) if you're going to change the spec, then there MUST be testing; and (4) this feels a lot like a change to an EIP without testing weeks for deployment.

What's the purpose of the EIP 1 process if it doesn't apply when we need it to not apply (clarifications included)?

**Comment by fubuloubu** [comment]
So, without this change merged, the EIP is still well-formed enough to deploy as is, as it is consistent with any currently known network that uses EIP-155 (all of which use less than 64/256 bits).

Perhaps the greater issue is that @axic identified a scenario with handling EIP-155 (and EIP-1344 by extension) where chain ID being unlimited in size would cause a logical error that could be disruptive to how RLP encoding is performed on transactions.

This modification (which should also be made to EIP-155) should actually be a follow-up EIP that can be implemented outside of the hardfork process since the mainnet and every other known network would not run into this scenario, but we want to make sure to bound this size so the issue is never encountered.

I will withdraw this amended change and communicate to the client teams that _have_ implemented this update so far to revert back to EIP-1344 returning a `uint256` instead of a `uint64`, and the assumption in EIP-155 hat chain ID has a potentially unlimited size.

**Comment by fubuloubu** [comment]
Got no response on this, so I am going to move ahead with the suggestion from @karalabe that this change proposal be dropped in favor of the implicit assumption from EIP-155 that chain ID is potentially "limitless" in size. This PR will therefore be dropped in favor of #2294, which would create this limit explicitly.

The original wording of EIP-1344 is still accurate and required because the EVM will return a 256-bit field in response to calling the `CHAIN_ID` opcode. If a value for Chain ID is chosen above this limit, it will be undefined behavior and could lead to potential security issues for chains that make that choice. This does not affect Ethereum, or known Ethereum-based chains, so the implementation of EIP-1344 should remain unaffected.

**Pull request closed without merging**